url888b6fd9b4_jmacglashan_commandsToTasks_tgz-pJ8-commands_tests_Model3TestsJ8/src/commands/model3/TaskModule.java

```diff
@@ -0,0 +1,893 @@
+				}
+			}
+			
+			List<GMQueryResult> res = new ArrayList<GMQueryResult>(TaskModule.this.liftedRFValues.size());
+			for(LiftedVarValue val : TaskModule.this.liftedRFValues){
+				Map<String, Integer> numOfEachObject = new HashMap<String, Integer>();
+				Set<String> seenVars = new HashSet<String>();
+				for(GroundedProp gp : val.conditions){
+					String [] paramClasses = gp.pf.getParameterClasses();
+					for(int i = 0; i < gp.params.length; i++){
+						if(!seenVars.contains(gp.params[i])){
+							//then it's a new variable
+							String cname = paramClasses[i];
+							Integer storedCount = numOfEachObject.get(cname);
+							int sc = storedCount != null ? storedCount : 0;
+							numOfEachObject.put(cname, sc++);
+						}
+					}
+				}
+				boolean possibleInState = true;
+				for(Entry<String, Integer> e : numOfEachObject.entrySet()){
+					if(e.getValue() > stateVal.s.getObjectsOfTrueClass(e.getKey()).size()){
+						possibleInState = false;
+						break;
+					}
+				}
+				
+				if(possibleInState){
+					GMQueryResult r = new GMQueryResult();
+					r.addQuery(val);
+					r.addCondition(stateVal);
+					res.add(r);
+				}
+				
+			}
+			
+			double uni = 1. / res.size();
+			for(GMQueryResult r : res){
+				r.probability = uni;
+			}
+			
+			this.sourceIterator = res.iterator();
+			
+		}
+		
+		@Override
+		public boolean hasNext() {
+			return this.sourceIterator.hasNext();
+		}
+
+		@Override
+		public void remove() {
+			throw new UnsupportedOperationException();
+		}
+
+		@Override
+		public GMQueryResult varSpecificNext() {
+			return this.sourceIterator.next();
+		}
+		
+		
+		
+	}
+	
+	
+	
+	
+	public class GroundedRFIterator extends ModelTrackedVarIterator{
+
+		protected StateRVValue				stateVal;
+		protected LiftedVarValue			liftedRFVal;
+		
+		protected List<RFConVariableValue>	rfVals;
+		protected int 						next = 0;
+		protected double					uniProb;
+		
+		public GroundedRFIterator(List<RVariableValue> conditions){
+			
+			for(RVariableValue val : conditions){
+				if(val.getOwner().equals(stateVariable)){
+					this.stateVal = (StateRVValue)val;
+				}
+				else if(val.getOwner().equals(liftedRFVariable)){
+					this.liftedRFVal = (LiftedVarValue)val;
+				}
+			}
+			
+			
+			String [] paramClasses = new String[liftedRFVal.freeVariableToClassMapping.size()];
+			String [] paramOrderGroups = new String[liftedRFVal.freeVariableToClassMapping.size()];
+			int i = 0;
+			for(Map.Entry<String, String> e : liftedRFVal.freeVariableToClassMapping.entrySet()){
+				paramClasses[i] = e.getValue();
+				paramOrderGroups[i] = e.getKey();
+				i++;
+			}
+			
+			List<List<String>> linearMappings = stateVal.s.getPossibleBindingsGivenParamOrderGroups(paramClasses, paramOrderGroups);
+			
+			List<Map<String, String>> mappings = new ArrayList<Map<String,String>>(linearMappings.size());
+			for(List<String> lmap : linearMappings){
+				Map<String, String> mapping = new HashMap<String, String>(lmap.size());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The `HashMap` constructor rehashes your hash table after 75% of all keys have been added to the table. To improve performance and prevent rehashing, use the Guava `newHashMapWithExpectedSize()` method. See the following diffs on GitHub that make this code change: [diff1](https://github.com/thefreebit/bullfrog/commit/d4d0b55af8a0eb925f039ef78d137565e1ae574b#diff-a24080844c93ebee0f466f384d2b3599cbfd502623597fc30e3364bc3fad5fedL119), [diff2](https://github.com/outbrain/ob1k/commit/5183bd681f4e9a8b67289314359b1211eb764899#diff-1891c9f8b2921c25e9406be78b9688662c7cd63b57300817fd288ed2ec0e3eeaL282)

Learn more: [HashMap constructor in the Java Platform Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html#HashMap-int-) and the [newHashMapWithExpectedSize in the Guava: Google Core Libraries for Java 23.0 API](https://guava.dev/releases/23.0/api/docs/com/google/common/collect/Maps.html#newHashMapWithExpectedSize-int-)